File: src/Snapfield.Core/Net/NetMessage.cs
using System.Buffers.Binary;
using System.Text.Json;
using System.Text.Json.Serialization;
using Snapfield.Core.Persistence;

namespace Snapfield.Core.Net;

public enum MsgType
{
    Hello,        // exchange machine id + monitors on connect
    CursorMove,   // warp receiver cursor to (X, Y) in its virtual-desktop pixels
    MouseButton,  // Button (0=L,1=R,2=M), Down
    MouseWheel,   // WheelDelta, Horizontal
    ControlEnter, // control just entered the receiver
    ControlLeave, // control just left the receiver
    Key,          // Vk, Scan, Down, Extended — keyboard transition
    Clipboard,      // Text — clipboard text changed on the sender
    AuthFail,       // receiver rejected the controller's pairing pin
    ClipboardImage, // Text carries base64 PNG — clipboard image changed on the sender
    Layout,         // Monitors — the controller's combined global plane, for the receiver to display
    ClipboardFiles, // Files — copied files (name + base64 bytes) shared via the clipboard
}

/// <summary>One file shared over the clipboard (name + base64-encoded content).</summary>
public sealed record SharedFile
{
    public string Name { get; init; } = "";
    public string Data { get; init; } = ""; // base64
}

/// <summary>
/// One wire message. A single flat record keeps (de)serialisation trivial; unused
/// fields stay at their defaults per message type. Frames are length-prefixed
/// (4-byte little-endian length + UTF-8 JSON body).
/// </summary>
public sealed record NetMessage
{
    public MsgType Type { get; init; }
    public string? MachineId { get; init; }
    public MonitorState[]? Monitors { get; init; }
    public int X { get; init; }
    public int Y { get; init; }
    public int Button { get; init; }
    public bool Down { get; init; }
    public int WheelDelta { get; init; }
    public bool Horizontal { get; init; }
    public int Vk { get; init; }
    public int Scan { get; init; }
    public bool Extended { get; init; }
    public string? Text { get; init; }
    public string? Pin { get; init; }
    public SharedFile[]? Files { get; init; }

    public static NetMessage Hello(string machineId, MonitorState[] monitors, string? pin = null) =>
        new() { Type = MsgType.Hello, MachineId = machineId, Monitors = monitors, Pin = pin };

    public static NetMessage ClipboardText(string text) => new() { Type = MsgType.Clipboard, Text = text };
    public static NetMessage ClipboardPng(byte[] png) => new() { Type = MsgType.ClipboardImage, Text = Convert.ToBase64String(png) };
    public static NetMessage AuthFailed() => new() { Type = MsgType.AuthFail };
    public static NetMessage LayoutSync(MonitorState[] monitors) => new() { Type = MsgType.Layout, Monitors = monitors };
    public static NetMessage ClipboardFilesMsg(SharedFile[] files) => new() { Type = MsgType.ClipboardFiles, Files = files };

    public static NetMessage Cursor(int x, int y) => new() { Type = MsgType.CursorMove, X = x, Y = y };
    public static NetMessage MouseBtn(int button, bool down) => new() { Type = MsgType.MouseButton, Button = button, Down = down };
    public static NetMessage Wheel(int delta, bool horizontal) => new() { Type = MsgType.MouseWheel, WheelDelta = delta, Horizontal = horizontal };
    public static NetMessage KeyEvent(int vk, int scan, bool down, bool extended) =>
        new() { Type = MsgType.Key, Vk = vk, Scan = scan, Down = down, Extended = extended };
    public static NetMessage Enter() => new() { Type = MsgType.ControlEnter };
    public static NetMessage Leave() => new() { Type = MsgType.ControlLeave };

    // ── binary fast-path ─────────────────────────────────────────────────────
    // Messages that flow at input rate (cursor at mouse-polling frequency, plus
    // buttons/wheel/keys) skip JSON entirely: a fixed little-endian layout
    // tagged by a first byte that can never start a JSON body ('{' = 0x7B), so
    // both encodings share the wire and FromBody picks by that byte.

    /// <summary>Upper bound of any binary-encoded body (see <see cref="TryEncodeBinary"/>).</summary>
    public const int MaxBinaryLength = 16;

    private const byte BinCursor = 0x01; // [tag][x:i32][y:i32]                 = 9
    private const byte BinButton = 0x02; // [tag][button:u8][down:u8]           = 3
    private const byte BinWheel  = 0x03; // [tag][delta:i32][horizontal:u8]     = 6
    private const byte BinKey    = 0x04; // [tag][vk:i32][scan:i32][flags:u8]   = 10

    /// <summary>Encodes a high-rate message into its fixed binary form. Returns
    /// false for message types that go as JSON. <paramref name="dest"/> must be
    /// at least <see cref="MaxBinaryLength"/> bytes.</summary>
    public bool TryEncodeBinary(Span<byte> dest, out int written)
    {
        switch (Type)
        {
            case MsgType.CursorMove:
                dest[0] = BinCursor;
                BinaryPrimitives.WriteInt32LittleEndian(dest[1..], X);
                BinaryPrimitives.WriteInt32LittleEndian(dest[5..], Y);
                written = 9;
                return true;
            case MsgType.MouseButton:
                dest[0] = BinButton;
                dest[1] = (byte)Button;
                dest[2] = Down ? (byte)1 : (byte)0;
                written = 3;
                return true;
            case MsgType.MouseWheel:
                dest[0] = BinWheel;
                BinaryPrimitives.WriteInt32LittleEndian(dest[1..], WheelDelta);
                dest[5] = Horizontal ? (byte)1 : (byte)0;
                written = 6;
                return true;
            case MsgType.Key:
                dest[0] = BinKey;
                BinaryPrimitives.WriteInt32LittleEndian(dest[1..], Vk);
                BinaryPrimitives.WriteInt32LittleEndian(dest[5..], Scan);
                dest[9] = (byte)((Down ? 1 : 0) | (Extended ? 2 : 0));
                written = 10;
                return true;
            default:
                written = 0;
                return false;
        }
    }

    /// <summary>Serialise the message body to UTF-8 JSON (no length prefix).</summary>
    public byte[] ToJson() => JsonSerializer.SerializeToUtf8Bytes(this, NetMessageJsonContext.Default.NetMessage);

    /// <summary>Serialise into a caller-owned writer — lets the send loop reuse
    /// its buffers instead of allocating per message.</summary>
    public void WriteTo(Utf8JsonWriter writer) =>
        JsonSerializer.Serialize(writer, this, NetMessageJsonContext.Default.NetMessage);

    /// <summary>Decodes a message body: binary fast-path frames by their tag
    /// byte, everything else as JSON. Null for malformed/unknown bodies.</summary>
    public static NetMessage? FromBody(ReadOnlySpan<byte> body)
    {
        if (body.Length == 0) return null;
        switch (body[0])
        {
            case BinCursor:
                if (body.Length != 9) return null;
                return Cursor(
                    BinaryPrimitives.ReadInt32LittleEndian(body[1..]),
                    BinaryPrimitives.ReadInt32LittleEndian(body[5..]));
            case BinButton:
                if (body.Length != 3) return null;
                return MouseBtn(body[1], body[2] != 0);
            case BinWheel:
                if (body.Length != 6) return null;
                return Wheel(BinaryPrimitives.ReadInt32LittleEndian(body[1..]), body[5] != 0);
            case BinKey:
                if (body.Length != 10) return null;
                return KeyEvent(
                    BinaryPrimitives.ReadInt32LittleEndian(body[1..]),
                    BinaryPrimitives.ReadInt32LittleEndian(body[5..]),
                    (body[9] & 1) != 0,
                    (body[9] & 2) != 0);
            default:
                try { return JsonSerializer.Deserialize(body, NetMessageJsonContext.Default.NetMessage); }
                catch (JsonException) { return null; }
        }
    }
}

// Omit default-valued fields: a JSON-encoded message carries only the fields it
// uses instead of all 15. Deserialising fills missing fields with the same
// defaults, so the round-trip is lossless. Source-generated: no reflection on
// the per-message path, and trimming/AOT safe.
[JsonSourceGenerationOptions(DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingDefault)]
[JsonSerializable(typeof(NetMessage))]
internal sealed partial class NetMessageJsonContext : JsonSerializerContext;

File: android/app/src/main/java/com/snapfield/receiver/net/NetMessage.kt
package com.snapfield.receiver.net

import org.json.JSONArray
import org.json.JSONObject

/**
 * Wire message — a 1:1 port of the desktop `NetMessage` record. The C# side
 * serialises with System.Text.Json defaults: PascalCase property names, enums
 * as integers, and default-valued fields omitted; this codec mirrors that
 * exactly so the two implementations are wire-compatible.
 */
object MsgType {
    const val Hello = 0
    const val CursorMove = 1
    const val MouseButton = 2
    const val MouseWheel = 3
    const val ControlEnter = 4
    const val ControlLeave = 5
    const val Key = 6
    const val Clipboard = 7
    const val AuthFail = 8
    const val ClipboardImage = 9
    const val Layout = 10
    const val ClipboardFiles = 11
}

/** One monitor as shipped in Hello/Layout — mirrors the desktop `MonitorState`. */
data class MonitorState(
    val machineId: String,
    val deviceId: String,
    val displayName: String,
    val pixelLeft: Int,
    val pixelTop: Int,
    val pixelWidth: Int,
    val pixelHeight: Int,
    val physicalXMm: Double,
    val physicalYMm: Double,
    val physicalWidthMm: Double,
    val physicalHeightMm: Double,
    val dpiScale: Double,
    val isInternal: Boolean,
    /** DeviceKind on the desktop: 0 unspecified, 1 monitor, 2 laptop, 3 phone, 4 tablet. */
    val kind: Int = 0,
) {
    fun toJson(): JSONObject = JSONObject().apply {
        put("MachineId", machineId)
        put("DeviceId", deviceId)
        put("DisplayName", displayName)
        if (pixelLeft != 0) put("PixelLeft", pixelLeft)
        if (pixelTop != 0) put("PixelTop", pixelTop)
        put("PixelWidth", pixelWidth)
        put("PixelHeight", pixelHeight)
        if (physicalXMm != 0.0) put("PhysicalXMm", physicalXMm)
        if (physicalYMm != 0.0) put("PhysicalYMm", physicalYMm)
        put("PhysicalWidthMm", physicalWidthMm)
        put("PhysicalHeightMm", physicalHeightMm)
        put("DpiScale", dpiScale)
        if (isInternal) put("IsInternal", true)
        if (kind != 0) put("Kind", kind)
    }
}

data class NetMessage(
    val type: Int = MsgType.Hello,
    val machineId: String? = null,
    val monitors: List<MonitorState>? = null,
    val x: Int = 0,
    val y: Int = 0,
    val button: Int = 0,
    val down: Boolean = false,
    val wheelDelta: Int = 0,
    val horizontal: Boolean = false,
    val vk: Int = 0,
    val scan: Int = 0,
    val extended: Boolean = false,
    val text: String? = null,
) {
    /** UTF-8 JSON body, defaults omitted — byte-compatible with the desktop. */
    fun toJsonBytes(): ByteArray {
        val o = JSONObject()
        if (type != 0) o.put("Type", type)
        machineId?.let { o.put("MachineId", it) }
        monitors?.let { list -> o.put("Monitors", JSONArray().apply { list.forEach { put(it.toJson()) } }) }
        if (x != 0) o.put("X", x)
        if (y != 0) o.put("Y", y)
        if (button != 0) o.put("Button", button)
        if (down) o.put("Down", true)
        if (wheelDelta != 0) o.put("WheelDelta", wheelDelta)
        if (horizontal) o.put("Horizontal", true)
        if (vk != 0) o.put("Vk", vk)
        if (scan != 0) o.put("Scan", scan)
        if (extended) o.put("Extended", true)
        text?.let { o.put("Text", it) }
        return o.toString().toByteArray(Charsets.UTF_8)
    }

    companion object {
        fun hello(machineId: String, monitors: List<MonitorState>) =
            NetMessage(type = MsgType.Hello, machineId = machineId, monitors = monitors)

        // Binary fast-path tags — mirror the desktop `NetMessage.TryEncodeBinary`
        // (v0.13.9+): input-rate messages skip JSON, tagged by a first byte that
        // can never start a JSON body ('{' = 0x7B).
        private const val BIN_CURSOR = 0x01 // [tag][x:i32 LE][y:i32 LE]        = 9
        private const val BIN_BUTTON = 0x02 // [tag][button:u8][down:u8]        = 3
        private const val BIN_WHEEL = 0x03  // [tag][delta:i32 LE][horizontal]  = 6
        private const val BIN_KEY = 0x04    // [tag][vk:i32 LE][scan:i32 LE][flags: bit0 down, bit1 extended] = 10

        private fun i32le(b: ByteArray, at: Int): Int =
            (b[at].toInt() and 0xFF) or
                ((b[at + 1].toInt() and 0xFF) shl 8) or
                ((b[at + 2].toInt() and 0xFF) shl 16) or
                ((b[at + 3].toInt() and 0xFF) shl 24)

        /** Decodes a body: binary fast-path frames by their tag byte, everything
        /// else as JSON — the exact mirror of the desktop `FromBody`. */
        fun fromBody(body: ByteArray, length: Int): NetMessage? {
            if (length == 0) return null
            when (body[0].toInt()) {
                BIN_CURSOR -> return if (length != 9) null else NetMessage(
                    type = MsgType.CursorMove, x = i32le(body, 1), y = i32le(body, 5))
                BIN_BUTTON -> return if (length != 3) null else NetMessage(
                    type = MsgType.MouseButton, button = body[1].toInt(), down = body[2].toInt() != 0)
                BIN_WHEEL -> return if (length != 6) null else NetMessage(
                    type = MsgType.MouseWheel, wheelDelta = i32le(body, 1), horizontal = body[5].toInt() != 0)
                BIN_KEY -> return if (length != 10) null else NetMessage(
                    type = MsgType.Key, vk = i32le(body, 1), scan = i32le(body, 5),
                    down = (body[9].toInt() and 1) != 0, extended = (body[9].toInt() and 2) != 0)
            }
            return fromJson(body, length)
        }

        private fun fromJson(body: ByteArray, length: Int): NetMessage? = try {
            val o = JSONObject(String(body, 0, length, Charsets.UTF_8))
            NetMessage(
                type = o.optInt("Type", 0),
                machineId = o.optString("MachineId").takeIf { o.has("MachineId") },
                monitors = null, // Layout/Hello monitor lists aren't consumed on Android yet
                x = o.optInt("X", 0),
                y = o.optInt("Y", 0),
                button = o.optInt("Button", 0),
                down = o.optBoolean("Down", false),
                wheelDelta = o.optInt("WheelDelta", 0),
                horizontal = o.optBoolean("Horizontal", false),
                vk = o.optInt("Vk", 0),
                scan = o.optInt("Scan", 0),
                extended = o.optBoolean("Extended", false),
                text = o.optString("Text").takeIf { o.has("Text") },
            )
        } catch (_: Exception) { null }
    }
}
